### PR TITLE
REQ-403 add fist to Cert_distrib.exchange_certificate_among_all_members

### PR DESCRIPTION
**Pull request by lippirk**
The idea here is that a test harness can place `/tmp/fist_exchange_certificates_among_all_members` on the file system of the primary host, and cause subsequent calls to`Pool.enable_tls_verification` to fail during its critical section. This will allow the test harness to mimic network failures, and reliably test their effect on this api call

### CODE REVIEW

File: ocaml/xapi/xapi_fist.ml [hunk 1]
```diff
@@ -122,3 +122,6 @@ let hang_psr psr_checkpoint =
       "psr_cleanup"
   )
   |> hang_while
+
+let exchange_certificates_among_all_members () =
```
Comment (lindig) [inline ocaml/xapi/xapi_fist.ml:126]: Would `exchange_certificates_in_pool` be clearer`? `all_members` makes me wonder if they are different from the pool.

**Comment by lippirk** [comment]
I made a couple of changes:

1. as requested, if you insert a fist point, a summary of what will happen is printed:

```
[debug||783 /var/lib/xcp/xapi|pool.enable_tls_verification D:c6023c9e0353|cert_distrib] exchange_certificates_among_all_members: we are about to...
[debug||783 /var/lib/xcp/xapi|pool.enable_tls_verification D:c6023c9e0353|cert_distrib] 0. send certs to b4cb481b
[debug||783 /var/lib/xcp/xapi|pool.enable_tls_verification D:c6023c9e0353|cert_distrib] 1. send certs to dbb4cb7a
[debug||783 /var/lib/xcp/xapi|pool.enable_tls_verification D:c6023c9e0353|cert_distrib] 2. FIST
```

2. when creating the fist file, you must add a seed. if you want to make a test reproducible, just use the same seed (and the same pool size)

File: ocaml/xapi/cert_distrib.ml [hunk 2]
```diff
@@ -372,21 +372,77 @@ let exchange_certificates_among_all_members ~__context =
...
+                  Api_errors.(
+                    Server_error
+                      ( internal_error
+                      , ["/tmp/fist_exchange_certificates_in_pool FIST!"]
+                      )
+                  )
+            )
+          in
+          let ops' =
+            List.fold_left
```
Comment (psafont) [inline ocaml/xapi/cert_distrib.ml:400]: Fold_left tends to make transformation on lists harder to inderstand. This looks like it has the same behaviour as `Xapi_stdext_std.Listext.List.take (rand_i - 1) @ [throw_op]`, this is not very efficient, adapting the code for `take` might be best here:
https://github.com/xapi-project/stdext/blob/master/lib/xapi-stdext-std/listext.ml#L61
```
  let take_and_append n op list =
    let rec loop i acc = function
      | x :: xs when (i - 1) < n ->
          loop (i + 1) (x :: acc) xs
      | _ ->
          List.rev acc
    in
    loop 0 [op] list
```

File: ocaml/xapi/cert_distrib.ml [hunk 3]
```diff
@@ -372,21 +372,77 @@ let exchange_certificates_among_all_members ~__context =
...
+                if i < rand_i then
+                  (i + 1, op :: acc)
+                else if i = rand_i then
+                  (i + 1, throw_op :: acc)
+                else
+                  (i + 1, acc)
+                )
+              (0, []) ops
+            |> snd
+            |> List.rev
```
Comment (psafont) [inline ocaml/xapi/cert_distrib.ml:411]: Fold_left tends to make transformation on lists harder to inderstand. This looks like it has the same behaviour as `Xapi_stdext_std.Listext.List.take (rand_i - 1) @ [throw_op]`, this is not very efficient, adapting the code for `take` might be best here:
https://github.com/xapi-project/stdext/blob/master/lib/xapi-stdext-std/listext.ml#L61
```suggestion
            let take_and_append n op list =
              let rec loop i acc = function
                | x :: xs when i < n ->
                    loop (i + 1) (x :: acc) xs
                | _ ->
                    List.rev acc
              in
              loop 1 [op] list
            in
            take_and_append rand_i throw_op ops
```

**Review by psafont — APPROVED** [review]

**Review by robhoes — APPROVED** [review]

**Comment by robhoes** [comment]
This looks good, but please squash the fixup before merging.

**Pull request merged**